### PR TITLE
Parse A Size Attribute 

### PR DESCRIPTION
**Pull request by sneha1302**
Hi , 

Have started the parse_a_size_attribute for image load conform , but need suggestions and so creating the pull Request .

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10827)

### CODE REVIEW

**Comment by jdm** [comment]
For the benefit of anyone else reading, copied from my last email:
```
Given that the sizes attribute follows a pattern (see example in http://bitsofco.de/the-srcset-and-sizes-attributes/ ), I recommend:
* define a structure that contains an optional media condition and a non-optional length
* make the function return instances of this new structure
* using Parser::try to attempt to parse a solitary length using Length::parse and returning that value if it succeeds
* otherwise, use MediaQuery::parse [1] to parse a media condition, then use Length::parse to parse a length, and return that value

You'll be using instances of the Result enum to properly interface with the parser, too: http://rustbyexample.com/error/result.html
```

If there are particular things which you are looking for suggestions about, please state them :)

**Comment by sneha1302** [comment]
@jdm : could you see if the function declaration looks right , and also - unparsed_size_list ? 
Thanks ! IF any further suggestion on how to go about white space token check , would be helpful !

**Comment by jdm** [comment]
I strongly recommend writing unit tests for your code so you have some idea of exactly what it's doing. You can do this directly in `tests/unit/script/lib.rs` by adding:
```rust
use script::dom::htmlimageelement::parse_a_sizes_attribute;
#[test]
fn some_parse_sizes_test() {
    let result = parse_a_sizes_attribute(...);
    assert!(/* something about result that is boolean */);
}
```
and run it with `./mach test-unit -p script`.

Now, as for the code: `expect_string` doesn't do what you want, which is presumably to return all of the non-comma characters in the string being parsed. I think the API of CSSParser is working against us here, and we should stop trying to use it in this code. Instead, we can use `let sizes = input.iter().split(',').collect::<Vec<_>>()` which will return a vector of strings and matches what you're attmpting to do in the following code.

As for the whitespace checks, we can use iterators to our advantage. The code looks like it should be equivalent to
```rust
let whitespace = unparsed_size.chars().rev().take_while(|c| util::str::char_is_whitespace(c)).count();
let trimmed = unparsed_size.chars().take(unparsed_size.chars().count() - whitespace).as_str();
```
`trimmed` will be a string that does not have any terminating whitespace after this.

**Comment by sneha1302** [comment]
The iterator was giving error , so included 
use std::iter::Iterator;

 let mut iter = Parser::new(input);
before iterating , but it gives error . What type should the iterator be , or look like?

**Comment by jdm** [comment]
We don't want to use `Parser::new` at all. We should be operating directly on the string, meaning the `input` argument to the function should be a `DOMString` value.

**Comment by sneha1302** [comment]
@jdm : 
for the second step in the for each unparsed_size (i.e If the last component value in unparsed size is a valid non-negative <source-size-value>,) . How do we go about checking if the value is valid or not  ? is there any existing function or method to do this check ? Your suggestions will be helpful 

Thanks !

**Comment by jdm** [comment]
Yes. Here's where we can use the cssparser::Parser API effectively - [Length::parse_non_negative](http://doc.servo.org/style/values/specified/enum.Length.html#method.parse_non_negative) accepts a reference to a `Parser` value, and returns a [Result](https://doc.rust-lang.org/nightly/book/error-handling.html#the-result-type) value that contains a `Length` if it was parsed correctly. This also correctly parses `calc` functions without permitting any others, based on my reading of `Length::parse_internal`. After parsing the value, you can use the [expect_whitespace](http://doc.servo.org/cssparser/struct.Parser.html#method.expect_whitespace) API, followed by [Expression::parse](http://doc.servo.org/style/media_queries/enum.Expression.html#method.parse) for the remaining steps.

**Comment by sneha1302** [comment]
@jdm
 `let trimmed = unparsed_size.chars().take(unparsed_size.chars().count() - whitespace).as_str();`
in the above code , as suggested . The as_str() method doesnot work as it says that "take" does not have that method . Any suggestion ?

**Comment by jdm** [comment]
My mistake for not trying it out in https://play.rust-lang.org/ first. This works:
```rust
let s = "foo bar   ".to_owned();
    let whitespace = s.chars().rev().take_while(|c| *c == ' ').count();
    let nonwhitespace: String = s.chars().take(s.chars().count() - whitespace).collect();
    println!("|{}|", nonwhitespace);
```

**Comment by sneha1302** [comment]
@jdm 
Could you check the code and suggest whats wrong and what all needs to be done . 

Thanks ! sneha

**Comment by jdm** [comment]
I strongly recommend writing unit tests that put your code through its paces. Start with the case of a single length value, then try two comma-separated values where the first one has a media expression, then three. Also try the cases where there's a parse error expected, and trailing whitespace, etc. http://bitsofco.de/the-srcset-and-sizes-attributes/ has a great example that shows some sample values that can exist in the `sizes` attribute.

File: components/script/dom/htmlimageelement.rs [hunk 1]
```diff
@@ -426,45 +427,45 @@ fn image_dimension_setter(element: &Element, attr: Atom, value: u32) {
...
+    }*/
+
+    //checking other media elements to obtain expression
     for unparsed_size in unparsed_sizes_list{
         let whitespace = unparsed_size.chars().rev().take_while(|c| util::str::char_is_whitespace(*c)).count();
-        let trimmed = unparsed_size.chars().take(unparsed_size.chars().count() - whitespace);
+        let trimmed: String = unparsed_size.chars().take(unparsed_size.chars().count() - whitespace).collect();
         
+        let mut parse_each = Parser::new(&trimmed);
+        let each_media_query = parse_each.try(MediaQuery::parse);
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: I'll point out that a media query is different than a media expression. We don't want to try to parse a whole media query here; I mentioned `Expression::parse` for a reason.

**Comment by srm912** [comment]
`Length:parse_non_negative` does not work for the value **calc(100%-3em)**. Works for **__em**

**Comment by srm912** [comment]
Also, shouldn't the `struct Size` encapsulate an optional `MediaQuery` and `Length` making it 
```
pub struct Size {
    query: Option<MediaQuery>,
    length: Length,
}
```

**Comment by srm912** [comment]
`MediaQuery::parse` does not recognize the following inputs:
```
( not (orientation: portrait) ) 300px,
( (orientation: landscape) or (min-width: 1000px) ) 50vw
```

**Comment by Ms2ger** [comment]
What happened here?

**Comment by srm912** [comment]
@Ms2ger This branch was under development. Since, it had some extra changes which had already been pushed to `servo:master` as part of #10767 , I removed that commit and reset all the new commits made on top of that one.

**Pull request merged**